jfix-stdlib-ratelimiter/src/test/kotlin/ru/fix/stdlib/ratelimiter/SuspendableRateLimitedDispatcherTest.kt

```diff
@@ -0,0 +1,713 @@
+        dispatch.close()
+
+        futures.forEach { future: CompletableFuture<*> ->
+            future.isDone.shouldBeTrue()
+            future.isCompletedExceptionally.shouldBeFalse()
+        }
+    }
+
+    /**
+     * When dispatcher closingTimeout is enough for pending tasks to complete
+     * such tasks will complete normally
+     */
+    @Test
+    @Timeout(10, unit = TimeUnit.SECONDS)
+    fun `on shutdown fast tasks complete normally`() = runBlocking {
+
+        val dispatch = createDispatcher(closingTimeout = 5_000)
+
+        val results = List(3) {
+            dispatch.decorateSuspend {
+                true
+            }
+        }
+        dispatch.close()
+
+        results.forEach { result ->
+            result.shouldBeTrue()
+        }
+    }
+
+    @Test
+    @Timeout(5, unit = TimeUnit.SECONDS)
+    fun `on shutdown slow tasks with CompletableFuture complete exceptionally`() {
+        val dispatch = createDispatcher(closingTimeout = 0)
+
+        for (i in 1..3) {
+            scope.async {
+                dispatch.compose {
+                    sleep(3000)
+                    completedFuture(true)
+                }
+            }
+        }
+
+        val exc = shouldThrow<RejectedExecutionException> {
+            dispatch.close()
+        }
+        exc.message!!.contains("timeout while waiting for coroutines finishing").shouldBeTrue()
+    }
+
+    @Test
+    @Timeout(5, unit = TimeUnit.SECONDS)
+    fun `on shutdown slow tasks complete exceptionally`() = runBlocking {
+        val dispatch = createDispatcher(closingTimeout = 0)
+
+        for (i in 1..3) {
+            scope.async {
+                dispatch.decorateSuspend {
+                    sleep(3000)
+                    true
+                }
+            }
+        }
+
+        val exc = shouldThrow<RejectedExecutionException> {
+            dispatch.close()
+        }
+        exc.message!!.contains("timeout while waiting for coroutines finishing").shouldBeTrue()
+    }
+
+    @Test
+    fun `task with CompletableFuture, submitted in closed dispatcher, is rejected with exception`() {
+        val dispatcher = createDispatcher()
+        dispatcher.close()
+
+        val t = shouldThrow<Exception> {
+            dispatcher.compose {
+                completedFuture(true)
+            }.join()
+        }
+        t.shouldBeInstanceOf<CompletionException>()
+        t.message!!.contains("TERMINATE")
+        return
+    }
+
+    @Test
+    fun `task, submitted in closed dispatcher, is rejected with exception`() = runBlocking {
+        val dispatcher = createDispatcher()
+        dispatcher.close()
+
+        val t = shouldThrow<Exception> {
+            dispatcher.decorateSuspend {
+                true
+            }
+        }
+        t.shouldBeInstanceOf<RejectedExecutionException>()
+        t.message!!.contains("TERMINATE")
+        return@runBlocking
+    }
+
+    @Test
+    fun `invoke suspend function wrapped to completable future`() {
```

Comment:
I think the things tested by this function are the same ones tested by `dispatch async operation with successfull CompletableFuture, operation invoked and it's result returned` function.

Thus I suggest to leave only this test which additionally check realyy suspendable function (not the one immediately returning result as in other test).

Reply:
Ok

Reply:
Removed